prov/util/src/util_cq.c

```diff
@@ -203,21 +203,35 @@ ssize_t ofi_cq_readerr(struct fid_cq *cq_fid, struct fi_cq_err_entry *buf,
 	struct util_cq *cq;
 	struct util_cq_err_entry *err;
 	struct slist_entry *entry;
-	ssize_t ret;
+	char *err_buf_save = NULL;
+	ssize_t ret = -FI_EAGAIN;
 
 	cq = container_of(cq_fid, struct util_cq, cq_fid);
 	fastlock_acquire(&cq->cq_lock);
-	if (!ofi_cirque_isempty(cq->cirq) &&
-	    (ofi_cirque_head(cq->cirq)->flags & UTIL_FLAG_ERROR)) {
-		ofi_cirque_discard(cq->cirq);
-		entry = slist_remove_head(&cq->err_list);
-		err = container_of(entry, struct util_cq_err_entry, list_entry);
-		*buf = err->err_entry;
-		free(err);
-		ret = 0;
-	} else {
-		ret = -FI_EAGAIN;
+	if (ofi_cirque_isempty(cq->cirq) ||
+		!(ofi_cirque_head(cq->cirq)->flags & UTIL_FLAG_ERROR))
+		goto unlock;
+
+	ofi_cirque_discard(cq->cirq);
+	entry = slist_remove_head(&cq->err_list);
+	err = container_of(entry, struct util_cq_err_entry, list_entry);
+	if (buf->err_data_size) {
+		if (buf->err_data_size < err->err_entry.err_data_size) {
+			ret = -FI_ETOOSMALL;
+			goto err1;
+		} else {
+			err_buf_save = buf->err_data;
+			memcpy(err_buf_save, err->err_entry.err_data,
+					err->err_entry.err_data_size);
+		}
 	}
+	*buf = err->err_entry;
```

Comment:
I guess, it can be done before `if (buf->err_data_size) {` if statement. it will help us to avoid lines that is below (229-230 lines), because we save address of `buf->err_data` here `err_buf_save = buf->err_data;` (i.e we can avoid `err_buf_save` variable)